new/property-defaults.md

```diff
@@ -0,0 +1,120 @@
+---
+RFC: unassigned
+Author: John Keiser <jkeiser@chef.io>
+Status: Draft
+Type: Standards Track
+---
+
+# Property Default Value Improvements
+
+This RFC addresses some longstanding issues with `default` values on properties (attributes), adds `reset_property` to Resource, and addresses how we will move core Chef resources over to use defaults and properties.
+
+## Motivation
+
+    As a Chef developer,
+    I want the core resources to use `property`,
+    So that people are cribbing off good examples.
+
+    As a Chef user,
+    I want the core resources not to actually change behavior,
+    So that I can rely on things I've built on top of them.
+
+## Specification
+
+### Non-Sticky Defaults
+
+Defaults are presently *sticky*, in that the first time a property is accessed, it is stored in the instance variable:
+
+```ruby
+class X < Chef::Resource
+  attribute :foo, default: lazy { Random.rand }
+end
+
+x 'blah' do
+  puts @foo #=> nil
+  puts foo  #=> 1
+  puts foo  #=> 1
+  puts @foo #=> 1
+end
+```
+
+Since this has some utility, we add a new keyword for computed defaults.
+
+#### Computed defaults
+
+```ruby
+class X < Chef::Resource
+  attribute :foo, computed: lazy { Random.rand }
+end
+
+x 'blah' do
+  puts @foo #=> nil
+  puts foo  #=> 1
+  puts foo  #=> 325
+  puts @foo #=> nil
+end
+```
+
+#### Frozen constants
+
+Sticky defaults create a serious issue for constants: *every instance* of a resource gets the same value.
+
+```ruby
+class X < Chef::Resource
+  attribute :foo, default: []
+end
+x 'a' do
+  foo << 1
+  foo << 2
+end
+x 'b' do
+  # ERROR we picked up the values that got added to 'a'
+  puts foo #=> [ 1, 2 ]
+end
+```
+
+To fix this, we propose freezing constant default values, and not assigning them to the instance.
+
+TODO an alternative is to dup them when assigning to an instance the way Chef::Config does.
```

Comment:
Eh, I've found it useful in a few cases where provider mixins mutate the resource in place. That's super niche though and I'm sure there are other ways to do it, but still not quite "always".